platform/wayland/cog-platform-wl.c

```diff
@@ -2125,108 +2339,44 @@ clear_egl (void)
-                                                 win_data.wl_surface,
-                                                 ZWP_FULLSCREEN_SHELL_V1_PRESENT_METHOD_DEFAULT,
-                                                 NULL);
-
-        /* Configure the surface so that it respects the width and height
-         * environment variables */
-        configure_surface_geometry(0, 0);
-    } else if (wl_data.shell != NULL) {
-        win_data.shell_surface = wl_shell_get_shell_surface(wl_data.shell, win_data.wl_surface);
-        g_assert(win_data.shell_surface);
-
-        wl_shell_surface_add_listener(win_data.shell_surface, &shell_surface_listener, 0);
-        wl_shell_surface_set_toplevel(win_data.shell_surface);
-
-        /* wl_shell needs an initial surface configuration. */
-        configure_surface_geometry(0, 0);
-    }
+    wl_surface_add_listener(win_data.shell_context->wl_shell_data.wl_surface, &surface_listener, NULL);
+    win_data.shell_context->functions.create_window();
 
     const char *env_var;
     if ((env_var = g_getenv("COG_PLATFORM_WL_VIEW_FULLSCREEN")) && g_ascii_strtoll(env_var, NULL, 10) > 0) {
         win_data.is_maximized = false;
         win_data.is_fullscreen = true;
-
-        if (wl_data.xdg_shell != NULL) {
-            xdg_toplevel_set_fullscreen(win_data.xdg_toplevel, NULL);
-        } else if (wl_data.fshell != NULL) {
-            win_data.should_resize_to_largest_output = true;
-            resize_to_largest_output();
-        } else if (wl_data.shell != NULL) {
-            wl_shell_surface_set_fullscreen(win_data.shell_surface, WL_SHELL_SURFACE_FULLSCREEN_METHOD_SCALE, 0, NULL);
-        } else {
-            g_warning("No available shell capable of fullscreening.");
-            win_data.is_fullscreen = false;
-        }
+        win_data.shell_context->functions.enter_fullscreen();
     } else if ((env_var = g_getenv("COG_PLATFORM_WL_VIEW_MAXIMIZE")) && g_ascii_strtoll(env_var, NULL, 10) > 0) {
         win_data.is_maximized = true;
         win_data.is_fullscreen = false;
-
-        if (wl_data.xdg_shell != NULL) {
-            xdg_toplevel_set_maximized (win_data.xdg_toplevel);
-        } else if (wl_data.shell != NULL) {
-            wl_shell_surface_set_maximized (win_data.shell_surface, NULL);
-        } else {
-            g_warning ("No available shell capable of maximizing.");
-            win_data.is_maximized = false;
-        }
+        win_data.shell_context->functions.maximize_surface();
     }
 
     return TRUE;
 }
 
 static void
-destroy_window (void)
+destroy_window(void)
 {
-    g_clear_pointer (&win_data.xdg_toplevel, xdg_toplevel_destroy);
-    g_clear_pointer (&win_data.xdg_surface, xdg_surface_destroy);
-    g_clear_pointer (&win_data.shell_surface, wl_shell_surface_destroy);
-    g_clear_pointer (&win_data.wl_surface, wl_surface_destroy);
+    g_clear_pointer(&win_data.shell_context->xdg_shell_data.xdg_toplevel, xdg_toplevel_destroy);
```

Comment:
`shell_context` is only going to be one type of shell, accessing it as multiple types like this almost certainly means you're calling one or more of these functions with invalid data. This should be factored out into a surface_destroy function in shell_functions.